### PR TITLE
[WIP] Introduce consensus code for Whisk (SSLE) with Curdleproofs

### PR DESCRIPTION
**Pull request by nalinbhardwaj**
Ethresearch post: https://ethresear.ch/t/whisk-a-practical-shuffle-based-ssle-protocol-for-ethereum/11763

Rebased version of #2800

cc @asn-d6

### CODE REVIEW

File: tests/core/pyspec/eth2spec/merge/mainnet.py [hunk 1]
```diff
@@ -0,0 +1,2931 @@
...
+
+    def assemble_block(self, block_hash: Hash32, timestamp: uint64, random: Bytes32) -> ExecutionPayload:
+        raise NotImplementedError("no default block production")
+
+
+EXECUTION_ENGINE = NoopExecutionEngine()
+
+
+assert FINALIZED_ROOT_INDEX == get_generalized_index(BeaconState, 'finalized_checkpoint', 'root')
+assert NEXT_SYNC_COMMITTEE_INDEX == get_generalized_index(BeaconState, 'next_sync_committee')
```
Comment (dapplion) [inline tests/core/pyspec/eth2spec/merge/mainnet.py:2931]: I don't think you want to commit this files. @hwwhww aren't this gitignored or what may have happened?

File: tests/core/pyspec/eth2spec/test/context.py [hunk 2]
```diff
@@ -8,18 +8,19 @@
 from eth2spec.bellatrix import mainnet as spec_bellatrix_mainnet, minimal as spec_bellatrix_minimal
 from eth2spec.capella import mainnet as spec_capella_mainnet, minimal as spec_capella_minimal
 from eth2spec.eip4844 import mainnet as spec_eip4844_mainnet, minimal as spec_eip4844_minimal
+from eth2spec.whisk import mainnet as spec_whisk_mainnet, minimal as spec_whisk_minimal
 from eth2spec.utils import bls
 
 from .exceptions import SkippedTest
 from .helpers.constants import (
-    PHASE0, ALTAIR, BELLATRIX, CAPELLA, EIP4844,
+    PHASE0, ALTAIR, BELLATRIX, CAPELLA, EIP4844, WHISK, SHARDING,
```
Comment (dapplion) [inline tests/core/pyspec/eth2spec/test/context.py:16]:
```suggestion
    PHASE0, ALTAIR, BELLATRIX, CAPELLA, EIP4844, WHISK,
```

File: tests/core/pyspec/eth2spec/test/helpers/constants.py [hunk 3]
```diff
@@ -15,13 +15,14 @@
...
 EIP4844 = SpecForkName('eip4844')
+WHISK = SpecForkName('whisk')
 
 # The forks that pytest can run with.
 ALL_PHASES = (
     # Formal forks
     PHASE0, ALTAIR, BELLATRIX, CAPELLA,
     # Experimental patches
-    EIP4844,
+    EIP4844, CUSTODY_GAME, WHISK,
```
Comment (dapplion) [inline tests/core/pyspec/eth2spec/test/helpers/constants.py:25]:
```suggestion
    EIP4844, WHISK,
```

File: configs/mainnet.yaml [hunk 4]
```diff
@@ -81,7 +77,6 @@ MIN_PER_EPOCH_CHURN_LIMIT: 4
 # 2**16 (= 65,536)
 CHURN_LIMIT_QUOTIENT: 65536
 
-
```
Comment (dapplion) [inline configs/mainnet.yaml:84]: Unnecessary polluting diff styling `configs/mainnet.yaml` and `configs/minimal.yaml`, should be reverted for a cleaner PR.

File: tests/core/pyspec/eth2spec/merge/mainnet.py (hunk 1 above)
Comment (hwwhww) [reply tests/core/pyspec/eth2spec/merge/mainnet.py:2931]: Right, it should be ignored. The spec name stub "merge" was very outdated. I guess @nalinbhardwaj didn't clean the old stub folder before pulling from the `dev` branch.

**Comment by dapplion** [comment]
Some notes regarding implementation:
- Instead of extending the validator class, add two new lists in the state. That reduces the additional hashing load from +4 hashes / val record to +2 hashes / val record. It also prevents adding 1 extra depth to the validator type and break light-client uses. It also simplifies implementation, as clients are used to different states but the Validator type has remained constant since phase0 and is heavily used.
- Move `whisk_opening_proof` to the block body, not in the BeaconBlock type. I see no reason to not add the property in the body, and similarly to comment above, adding new properties outside the body adds implementation complexity for no clear gain.
- Use strict snake case, Lodestar does not expect capital letters after `_` otherwise it would break tooling.

See suggested diff here https://github.com/nalinbhardwaj/consensus-specs/compare/whisk-curdleproofs...dapplion:consensus-specs:whisk-curdleproofs-review

---

If you want @nalinbhardwaj I can help rebase this PR into current structure, as now features are in a different directory than when this PR was created.

**Comment by asn-d6** [comment]
> 



> Some notes regarding implementation:
> 
>     * Instead of extending the validator class, add two new lists in the state. That reduces the additional hashing load from +4 hashes / val record to +2 hashes / val record. It also prevents adding 1 extra depth to the validator type and break light-client uses. It also simplifies implementation, as clients are used to different states but the Validator type has remained constant since phase0 and is heavily used.
> 
>     * Add `whisk_opening_proof` in the block body, not in the BeaconBlock type. I see no reason to not add the property in the body, and similarly to comment above, adding new properties outside the body adds implementation complexity for no clear gain.
> 
>     * Use strict snake case, Lodestar does not expect capital letters after `_` otherwise it would break tooling.
> 
> 
> See suggested diff here [nalinbhardwaj/consensus-specs@whisk-curdleproofs...dapplion:consensus-specs:whisk-curdleproofs-review](https://github.com/nalinbhardwaj/consensus-specs/compare/whisk-curdleproofs...dapplion:consensus-specs:whisk-curdleproofs-review)
> 
> If you want @nalinbhardwaj I can help rebase this PR into current structure, as now features are in a different directory than when this PR was created.

Hey @dapplion . Thanks for reviewing the code. That was invaluable.

I think all your suggestions are great with regards to better utilizing the state and the beacon blocks.

I don't have a strong opinion about the naming, but I kinda liked having capital letters to signify the group elements (instead of field elements). Instead of changing the spec, isn't this something that can be changed on the lodestar side?

With regards to incorporating your diff: Would you like to rebase this PR, and then do a PR on top of it, so that we are also able to review your changes? I think most of your changes are great, but perhaps I'd like to comment on a line or two.

**Comment by hwwhww** [comment]
Hey, @nalinbhardwaj  @asn-d6 

Could you move the feature specs to `specs/_features/whisk`?

I suppose it needs more proper reviews before merging, but I can at least add a CI job to filter out the basic bugs. :)

**Comment by hwwhww** [comment]
Closing via #3342

Thank you @nalinbhardwaj @asn-d6 @JustinDrake @dapplion!

**Pull request closed without merging**